lib/Performance/Probability.pm

```diff
@@ -129,54 +129,143 @@ sub _variance_x_square {
-                    my $second_contract_duration = ($sell_time->[$j] - $start_time->[$j]);
+                        #calculate first and second contracts durations. please see the documentation for details
 
-                    my $i_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$i]);
-                    my $j_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$j]);
+                        my $first_contract_duration  = ($sell_time->[$i] - $start_time->[$i]);
+                        my $second_contract_duration = ($sell_time->[$j] - $start_time->[$j]);
 
-                    my $corr_ij = $b_interval / (sqrt($first_contract_duration) * sqrt($second_contract_duration));
+                        my $i_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$i]);
+                        my $j_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$j]);
 
-                    if ($types->[$i] ne $types->[$j]) {
-                        $corr_ij = -1 * $corr_ij;
-                    }
+                        my $corr_ij = $b_interval / (sqrt($first_contract_duration) * sqrt($second_contract_duration));
 
-                    if ($corr_ij < -1 or $corr_ij > 1) {
-                        next;
-                    }
+                        if ($types->[$i] ne $types->[$j]) {
+                            $corr_ij = -1 * $corr_ij;
+                        }
+
+                        if ($corr_ij < -1 or $corr_ij > 1) {
+                            next;
+                        }
+
+                        my $p_ij = Math::BivariateCDF::bivnor($i_strike, $j_strike, $corr_ij);
 
-                    my $p_ij = Math::BivariateCDF::bivnor($i_strike, $j_strike, $corr_ij);
+                        my $covariance_ij =
+                            ($p_ij - $pk->[$i] * $pk->[$j]) * ($wk->[$i] - $lk->[$i]) * ($wk->[$j] - $lk->[$j]);
+                        $covariance = $covariance + $covariance_ij;
+                    }
+                }
+            }
+        } elsif ($types->[$i] =~ /^DIGIT/) {
+            for ($j = 0; $j < @{$exit_tick_epoch}; ++$j) {
+                if ($i != $j and $underlying->[$i] eq $underlying->[$j] and $exit_tick_epoch->[$i] == $exit_tick_epoch->[$j]) {
 
-                    my $covariance_ij =
-                        ($p_ij - $pk->[$i] * $pk->[$j]) * ($wk->[$i] - $lk->[$i]) * ($wk->[$j] - $lk->[$j]);
+                    my $p_ij = get_shared_winning_probability({
+                            type_1    => $types->[$i],
+                            type_2    => $types->[$j],
+                            barrier_1 => $barriers->[$i],
+                            barrier_2 => $barriers->[$j]});
 
+                    my $covariance_ij = ($p_ij - $pk->[$i] * $pk->[$j]) * ($wk->[$i] - $lk->[$i]) * ($wk->[$j] - $lk->[$j]);
                     $covariance = $covariance + $covariance_ij;
+
                 }
             }
+        } else {
+
+            next;
```

Comment:
maybe instead of silently dropping we should `warn`?

Reply:
why we want to warn ? We are not going to include any contracts apart from call put and digit into performance probability calculation. It is intended to drop them